gnomad_constraint/utils/constraint.py

```diff
@@ -386,3 +392,171 @@ def apply_models(
+
+    .. note::
+        The following annotations should be present in `ht`:
+            - modifier
+            - annotation
+            - observed_variants
+            - mu
+            - possible_variants
+            - expected_variants
+            - expected_variants_{pop} (if `pops` is specified)
+            - downsampling_counts_{pop} (if `pops` is specified)
+
+    :param ht: Input Table with the number of expected variants (output of
+        `get_proportion_observed()`).
+    :param keys: The keys of the output Table, defaults to ('gene', 'transcript',
+        'canonical').
+    :param classic_lof_annotations: Classic LoF Annotations used to filter the input
+        Table. Default is {"stop_gained", "splice_donor_variant",
+        "splice_acceptor_variant"}.
+    :param pops: List of populations used to compute constraint metrics. Default is ().
+    :param expected_values: Dictionary containing the expected values for 'Null',
+        'Rec', and 'LI' to use as starting values.
+    :param min_diff_convergence: Minimum iteration change in LI to consider the EM
+        model convergence criteria as met. Default is 0.001.
+    :param raw_z_outlier_threshold: Value at which the raw z-score is considered an outlier. Values below the negative of '--raw-z-outlier-threshold' will be considered outliers for lof and missense varaint counts (indicating too many variants), whereas values either above '--raw-z-outlier-threshold' or below the negative of '--raw-z-outlier-threshold' will be considered outliers for synonymous varaint counts (indicating too few or too many variants)."
+    :return: Table with pLI scores, observed:expected ratio, confidence interval of the
+        observed:expected ratio, and z scores.
+    """
+    classic_lof_annotations = hl.literal(classic_lof_annotations)
+    # This function aggregates over genes in all cases, as XG spans PAR and non-PAR X.
+    # `annotation_dict` stats the rule of filtration for each annotation.
+    annotation_dict = {
+        # Filter to classic LoF annotations with LOFTEE HC or LC.
+        "lof_hc_lc": classic_lof_annotations.contains(ht.annotation)
+        & ((ht.modifier == "HC") | (ht.modifier == "LC")),
+        # Filter to LoF annotations with LOFTEE HC or OS.
+        "lof_hc_os": (ht.modifier == "HC") | (ht.modifier == "OS"),
+        # Filter to LoF annotations with LOFTEE HC.
+        "lof": ht.modifier == "HC",
+        # Filter to missense variants.
+        "mis": ht.annotation == "missense_variant",
+        # Filter to probably damaging missense variants predicted by PolyPen-2.
+        "mis_pphen": ht.modifier == "probably_damaging",
+        # Filter to synonymous variants.
+        "syn": ht.annotation == "synonymous_variant",
+    }
+
+    # Compute the observed:expected ratio. Will not compute per pop for "mis_pphen".
```

Comment:
I changed some things around in the code below. I was having some issues with the flags specific to each type where I couldn't add to the struct if it was missing, which is the case for some of the flags. Instead I changed to only have the one flag annotation at the end, but specify which to use for filtering in the sd computation